zips/zip-0231.md

```diff
@@ -0,0 +1,332 @@
+  one for each subsequent chunk within a particular memo.
+- `final_chunk` is the byte `0x01` for the final memo chunk, and `0x00` for all
+  preceding chunks.
+
+Finally, the encrypted memo chunks for all memos are combined into a single
+sequence using an order-preserving shuffle. Memo chunks from different memos MAY
+be interleaved in any order, but memo chunks from the same memo MUST have the
+same relative order. The following diagram shows an example shuffle of three
+memos:
+
+```
+[
+    (memo_a, 0),
+    (memo_b, 0),
+    (memo_a, 1),
+    (memo_c, 0),
+    (memo_c, 1),
+    (memo_a, 2),
+]
+```
+
+## Memo decryption
+
+When a recipient decrypts a shielded output, they obtain a `memo_key`. From this
+they derive `encryption_key` as above, and then proceed as follows:
+
+- Set `counter = 0` and `final_chunk = 0x00`.
+- Attempt to decrypt each memo chunk in order. Pruned memo chunks are skipped.
+- Each time decryption succeeds for a memo chunk, increment `counter` by 1, and
+  then continue attempting to decrypt subsequent chunks.
+- Once all memo chunks have been trial-decrypted once, set `final_chunk = 0x01`
+  and then attempt to decrypt the memo chunks again, starting immediately after
+  the last successfully-decrypted chunk (or at the start if none were), and
+  without incrementing `counter`.
+  - This step can be made secret-independent by attempting to decrypt every memo
+    chunk again, and ignoring the results of all chunks up to and including the
+    last successfully-decrypted chunk.
+- If no memo chunk decrypts successfully with `final_chunk = 0x01`, discard any
+  memo chunks that were decrypted, and return nothing. Otherwise, concatenate
+  the decrypted memo chunks in order and return the concatenation as the memo.
+
+If any chunk of the memo encrypted to `memo_key` has been pruned, the decryption
+process above returns nothing (as `final_chunk` will be set to `0x01` with the
+wrong counter value), ensuring that a malformed memo is not returned.
+
+## Encoding in transactions
+
+| Bytes  | Name          | Data Type                | Description                             |
+|--------|---------------|--------------------------|---------------------------------------- |
+| 32     | `nonce`       | `byte[32]`               | The nonce for deriving encryption keys. |
+| 8      | `pruned`      | `uint64`                 | Bitflags indicating the type of each entry in `vMemoChunks`. Bits are little endian. 0 corresponds to `byte[272]` and is an encrypted memo chunk. 1 corresponds to `byte[32]` and is the `memo_chunk_digest` for a pruned chunk. |
+| varies | `nMemoChunks` | `compactSize`            | Length of `vMemoChunks`.                |
+| varies | `vMemoChunks` | `MemoChunk[nMemoChunks]` | A sequence of encrypted memo chunks.    |
+
+TODO: Decide whether to permit an optimised encoding for a fully-pruned memo
+bundle that just contains `memo_bundle_digest`.
```

Comment:
In thinking about and discussing pruning with @str4d and @arya2, several concerns arose related to pruning.

* If pruning is permitted in the network protocol format, what does the behavior of nodes need to be when they receive a pruned transaction? Do they continue attempting to retrieve the full transaction data from other peers? If so, does this create a potential situation where nodes have to keep track of which peers they've re-requested transactions from when attempting to restore complete memo data? How long do nodes need to continue retrying in order to ensure that they have obtained all the memo data that is available?
* Does pruning introduce a risk to the usability of memos on the network, due to a tragedy of the commons where node operators prefer to prune memos to save on storage and bandwidth, and as a consequence data availability suffers?
* Is there a potential adversarial case to consider, where some party cheaply runs a bunch of nodes on the network, all of which prune memo data when propagating transactions, thereby directly attacking data availability?

It's my personal opinion that the risks and complexity introduced by the pruning functionality make it infeasible to implement pruning in NU7. It might be something that can be introduced at a later date, but given the other functionality that we're currently aiming to support in NU7, we should attempt to avoid any changes that impact the network protocol if at all possible.